### PR TITLE
Support decompositions from numpy.linalg

### PR DESCRIPTION
**Pull request by pitrou**
Adds support for:
 * np.linalg.cholesky()
 * np.linalg.eig()
 * np.linalg.svd()

Along with tests.

Additional changes:
 * support for np.copy()
 * support for np.asfortranarray()
 * support for casting the `ctypes` attribute of an array to the corresponding raw pointer (for example to `types.CPointer(types.double)` for a float64 array)
 * support for casting the `ctypes` attribute of an array to `types.voidptr`

NOTE(s):
 * The function eig() will only return a real space result from
   a real space input, it will raise an exception if complex
   eigenvalues are encountered.
 * The function svd() only supports the kwarg "full_matrices"
   and will by always act as though "compute_uv" is set to 1.
 * All the functions above work in single dispatch mode only,
   they are not gufuncs.

### CODE REVIEW

**Comment by stuartarchibald** [comment]
@pitrou any idea why this fails here: https://travis-ci.org/numba/numba/jobs/124492254#L671 ?

**Comment by pitrou** [comment]
Perhaps `asfortranarray` behaves differently on Numpy 1.7. I'll have to take a look.

File: numba/typing/npydecl.py [hunk 1]
```diff
@@ -775,23 +798,38 @@ def generic(self, args, kws):
...
-                raise TypingError("np.linalg.inv() only supported on "
-                                  "float and complex arrays")
+            _check_linalg_matrix(a, "inv")
             return a.copy(layout='C')
 
         return typer
 
 
+@infer_global(numpy.linalg.qr)
+class LinalgQR(CallableTemplate):
```
Comment (stuartarchibald) [inline numba/typing/npydecl.py]: We can probably remove this. Chances are it'll be implemented with `@overload`.

File: numba/_lapack.c [hunk 2]
```diff
@@ -751,26 +751,39 @@ numba_ez_cgeev(char kind, char jobvl, char jobvr,  Py_ssize_t n, void *a,
     return ez_geev_return(info);
 }
 
+
+static int
+ez_gesdd_return(int info)
```
Comment (stuartarchibald) [inline numba/_lapack.c]: Should be `F_INT` ?

Comment (pitrou) [inline numba/_lapack.c]: Actually, should be Py_ssize_t according to other declarations :)

**Comment by sklam** [comment]
In python 2.7, there are many warnings in during `setup.py build_ext` and test_linalg.py fails with:

```bash
test_linalg_cholesky (__main__.TestLinalgCholesky) ... ok
test_linalg_eig (__main__.TestLinalgEig) ... FAIL
test_linalg_inv (__main__.TestLinalgInv) ... ok
test_linalg_svd (__main__.TestLinalgSvd) ... FAIL
test_contiguity_warnings (__main__.TestProduct) ... ok
test_dot_mm (__main__.TestProduct) ... ok
test_dot_vm (__main__.TestProduct) ... ok
test_dot_vv (__main__.TestProduct) ... ok
test_matmul_mm (__main__.TestProduct) ... skipped 'the matrix multiplication operator needs Python 3.5+ and Numpy 1.10+'
test_matmul_vm (__main__.TestProduct) ... skipped 'the matrix multiplication operator needs Python 3.5+ and Numpy 1.10+'
test_matmul_vv (__main__.TestProduct) ... skipped 'the matrix multiplication operator needs Python 3.5+ and Numpy 1.10+'
test_vdot (__main__.TestProduct) ... ok

======================================================================
FAIL: test_linalg_eig (__main__.TestLinalgEig)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/Users/sklam/dev/numba/numba/tests/test_linalg.py", line 568, in test_linalg_eig
    check(a)
  File "/Users/sklam/dev/numba/numba/tests/test_linalg.py", line 564, in check
    del got, expected
  File "/Users/sklam/dev/.conda.workspace/numba_devel27/lib/python2.7/contextlib.py", line 24, in __exit__
    self.gen.next()
  File "numba/tests/support.py", line 130, in assertNoNRTLeak
    "number of data allocs != number of data frees")
AssertionError: 6L != 5L : number of data allocs != number of data frees

======================================================================
FAIL: test_linalg_svd (__main__.TestLinalgSvd)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/Users/sklam/dev/numba/numba/tests/test_linalg.py", line 704, in test_linalg_svd
    check(a, full_matrices=fmat)
  File "/Users/sklam/dev/numba/numba/tests/test_linalg.py", line 690, in check
    del got, expected
  File "/Users/sklam/dev/.conda.workspace/numba_devel27/lib/python2.7/contextlib.py", line 24, in __exit__
    self.gen.next()
  File "numba/tests/support.py", line 130, in assertNoNRTLeak
    "number of data allocs != number of data frees")
AssertionError: 5L != 4L : number of data allocs != number of data frees

----------------------------------------------------------------------
Ran 12 tests in 13.553s

FAILED (failures=2, skipped=3)
```

**Comment by pitrou** [comment]
> In python 2.7, there are many warnings in during setup.py build_ext and test_linalg.py fails with:

Is that on your own machine? Travis seems fine.

**Comment by pitrou** [comment]
> In python 2.7, there are many warnings in during setup.py build_ext and test_linalg.py fails with:

Ok, I can reproduce here. I'm pushing a fix for the failures and the build warnings.

File: numba/tests/test_linalg.py [hunk 3]
```diff
@@ -640,54 +641,55 @@ def test_linalg_svd(self):
...
+                np.testing.assert_allclose(
+                    a,
+                    rec,
+                    rtol=10 * resolution,
+                    atol=100 * resolution  # zeros tend to be fuzzy
+                )
+
+            # Ensure proper resource management
+            with self.assertNoNRTLeak():
+                cfunc(a)
```
Comment (sklam) [inline numba/tests/test_linalg.py]: add ``**kwargs`` to match usage on line 645 ?

Comment (pitrou) [inline numba/tests/test_linalg.py]: Good catch.

**Comment by stuartarchibald** [comment]
Thanks for the review and merge.

**Pull request merged**